### PR TITLE
blackbox OpenLog and serial_RX on same UART

### PR DESCRIPTION
**Pull request by Pierre-A**
Serial_RX receiver and blackbox OpenLog use RX and TX lines of UART respectively.
Up to now they cannot share the same UART which results in costly configuration.

This proposal allows connecting a serial RX and an OpenLog board on the same UART through full duplex operation. The RX and TX communications are here limited to 115200 bds (hardcoded serialRX baudrate).

I only tested the code with a sumD receiver and a 3500us looptime. Any testings for other receivers (spektrum, sbus, sumh and xbus) and looptime are of course welcome.

This mod has no incidence on recording data to the onboard flash memory chip.

Thanks for your comments.

### CODE REVIEW

File: src/main/io/serial.c [hunk 1]
```diff
@@ -200,7 +200,8 @@ bool isSerialConfigValid(serialConfig_t *serialConfigToCheck)
     for (index = 0; index < SERIAL_PORT_COUNT; index++) {
         serialPortConfig_t *portConfig = &serialConfigToCheck->portConfigs[index];
 
-        if (portConfig->functionMask & FUNCTION_MSP) {
+        if ((portConfig->functionMask & FUNCTION_MSP)
```
Comment (ledvinap) [inline src/main/io/serial.c]: MSP is counted here (one must be available)

File: src/main/io/serial.c [hunk 2]
```diff
@@ -211,11 +212,13 @@ bool isSerialConfigValid(serialConfig_t *serialConfigToCheck)
                 return false;
             }
 
-            if (!(portConfig->functionMask & FUNCTION_MSP)) {
```
Comment (ledvinap) [inline src/main/io/serial.c]: You must handle 2 cases here:
- MSP is set, check second function
- Blackbox is set, check second function (serialrx)

Also only some type of serial RX is shareable

File: src/main/rx/sumd.c [hunk 3]
```diff
@@ -53,17 +53,29 @@ bool sumdInit(rxConfig_t *rxConfig, rxRuntimeConfig_t *rxRuntimeConfig, rcReadRa
...
+    	if (portConfig->functionMask & ALL_FUNCTIONS_SHARABLE_WITH_RX_SERIAL)
+            serialPortMode = MODE_RXTX;
+    	else
+    		return false;
+    	}
+    else {
+        serialPortMode = MODE_RX;
+    }
+
+    serialPort_t *sumdPort = openSerialPort(portConfig->identifier, FUNCTION_RX_SERIAL, sumdDataReceive, SUMD_BAUDRATE, serialPortMode, SERIAL_NOT_INVERTED);
```
Comment (ledvinap) [inline src/main/rx/sumd.c]: MODE_INVERTED will probably clash with blackbox (it is possible to set inversion for each direction, but it is not implemented yet)

File: src/main/io/serial.c (hunk 2 above)
Comment (Pierre-A) [inline src/main/io/serial.c]: I'm not sure I understand what you mean as I only need to check this specific combination in case of bitCount > 1.
See further, the port sharing is also filtered in blackbox_io.c.

Comment (ledvinap) [inline src/main/io/serial.c]: FUNCTION_MSP | FUNCTION_RX_SERIAL gets accepted ...

Comment (Pierre-A) [inline src/main/io/serial.c]: You're right, but this is not related to the blackbox-serialRx sharing. 
This case (*FUNCTION_MSP | FUNCTION_RX_SERIAL*) is actually accepted but not tested as others are (like telemetry). Should add a *FUNCTION_RX_SERIAL*  in *ALL_FUNCTIONS_SHARABLE_WITH_MSP* (although tested in *ALL_FUNCTIONS_SHARABLE_WITH_BLACKBOX*)

Comment (ledvinap) [inline src/main/io/serial.c]:
```
if(portConfig->functionMask & FUNCTION_MSP)
   return portConfig->functionMask & ALL_FUNCTIONS_SHARABLE_WITH_MSP;
if(portConfig->functionMask & FUNCTION_BLACKBOX)
  return  portConfig->functionMask & ALL_FUNCTIONS_SHARABLE_WITH_BLACKBOX;
return false;
```

File: src/main/rx/sumd.c (hunk 3 above)
Comment (Pierre-A) [inline src/main/rx/sumd.c]: Yes, for now it's hardcoded as *not_inverted* for serial receivers except sbus which should need an external inverter to be considered as not_inverted.

Comment (ledvinap) [inline src/main/rx/sumd.c]: Sorry, commented wrong line, should be at SBUS code. SBUS_PORT_OPTIONS is inverted, TX will be inverted too on 303

**Comment by sprazlzapfn** [comment]
Great to find this Pull-Request!! I thought about submitting a feature request on this but was not sure if the one UART supports different baud rates for Rx (S-Bus) and Tx (Openlog). I take from your changes that this is the case also on 103 hardware, right?

**Comment by ledvinap** [comment]
The speed is identical for both directions, to openLog must be configured accordingly

**Comment by Pierre-A** [comment]
sbus baudrate is hardcoded to 100000 bds so you should specify this value in your blackbox *CONFIG* file.

**Comment by sprazlzapfn** [comment]
Hmm, thanks for your feedback! Using relatively fast loop-times (~1500) I encountered dropped data until I upped the rate to 250 kbaud. Need to check again. Probably have to use blackbox_rate to restrict the bandwidth.

**Comment by Pierre-A** [comment]
**@sprazizapfn**, thanks for reporting. Your 250kbds OpenLog rate will be ignored anyway as the baudrate is fixed by the serial_Rx one. When testing with sumD i tried 2500 us as looptime and had no dropout. Did not take time for testing shorter.

**Comment by ledvinap** [comment]
Technically it may be possible to use pause in sbus data to transmit blackbox at high speed.
Just reconfigure port to very high speed when SBUS frame ends and transmit block that is guarantied to fit in SBUS pause. And setup SBUS speed before next frame begins.

But is it worth it? serial ports are quite abundant on new controllers ...

**Comment by Pierre-A** [comment]
Just tried a 1500us looptime with a 1/2 blackbox ratio, no dropped data at all. So probably no need to make things more complicate.

**Comment by joshuabardwell** [comment]
>sbus baudrate is hardcoded to 100000 bds so you should specify this value in your blackbox CONFIG file.

This may be a deal-breaker for anybody running Betaflight, which forces 1000 looptime for gyro sync. Even with 1:2 logging, I get baud rates around 130 to 150 kBaud. I can't do 1:1 logging at all. 2000 or 3000 us sampling period is plenty fast enough to troubleshoot and tune, but you get aliasing signals that show up in your traces.

But honestly, it's good to just give people the option, so maybe carry on.

**Comment by hydra** [comment]
I do think there is some merit in this feature but supporting and testing the code, handling the edge cases (inversion, etc) documenting it and supporting the users of it I'm not sure it's really worth the development effort.  I feel @ledvinap is right, newer targets have more IO.

I'm going to mark this is low-priority for now.  If people want this feature please add a '+1' comment to this issue and we'll see if it's popular enough to warrant development time.

**Comment by Pierre-A** [comment]
No problem, I coded it for one "not_enough_ports" config and just wanted to share the code .. just in case...

**Comment by acc-** [comment]
I developed something similar: sharing SerialRX + FrSky telemetry (with minor changes it would be possible to add Blackbox as well). Also for "not enough ports" receiver, that is naze32 :)

A quick description:
- UART is initialized during SerialRX init (but MODE_RX is changed to MODE_RXTX), so generally SerialRX is the master over telemetry
- FrSky code reuses the serialport instead of opening it
- FrSky is fooled to have PORTSHARING_NOT_SHARED

This feature is quite useful in simple OpenLRSng based receivers. Original FrSky runs at 9600 baud, in this scenario it runs at the same speed as SerialRX. But it's not a problem for OpenLRSng, because once using serial rx mode, it still accepts telemetry data at serial rx speed. So OpenLRSng can read telemetry data at eg 115200 and transfer it.

I've tested it with naze32 - uart2 for SUMD and FrSky telemetry set, then it goes to my μLRS receiver. I can see the telemetry data back on my Taranis.

Attaching a patch to original cleanflight code instead of pull request, might be useful for you.

Here it is: [frsky-serial_rx-patch.zip](https://github.com/cleanflight/cleanflight/files/93524/frsky-serial_rx-patch.zip)

Or maybe it should be the separate pull request...
